### PR TITLE
MediaRecorder: Add media recorder manual test page to testrtc.

### PR DESCRIPTION
**Pull request by cricdecyan**
**Description**
This CL adds a manual test for Media Stream Recorder

**Purpose**
To be able to manually set the different parameters of a media recorder and record a local/remote peer connection stream

### CODE REVIEW

File: src/manual/media-recorder/js/main.js [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+    this.onButtonClick('Stop', this.stopRecorder.bind(this));
+    this.onButtonClick('RequestData', this.requestDataRecorder.bind(this));
+    this.onButtonClick('Playback', this.playback.bind(this));
+    this.onButtonClick('Download', this.download.bind(this));
+    var monitorSelect = document.getElementById('monitorEvents');
+    monitorSelect.onchange = this.monitorRecorderEvents.bind(this, event);
+  },
+
+  // This function creates a button onClick event handler.
+  onButtonClick: function(buttonName, callBack) {
```
Comment (KaptenJansson) [inline src/manual/media-recorder/js/main.js]: Not a fan of wrapping the event listener (or any standard method) method like this.

What about keeping it simple albeit more lines of codes:
```
var getUserMedia = document.querySelector('#GetUserMedia');
getUserMedia.addEventListener('click', this.getUserMedia.bind(this));
```

File: src/manual/media-recorder/js/main.js [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+    console.log('button name', buttonName);
```
Comment (KaptenJansson) [inline src/manual/media-recorder/js/main.js]: Remove debug logging.

File: src/manual/media-recorder/js/main.js [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+    document.getElementsByName(buttonName)[0].addEventListener('click',
+        callBack);
+  },
+
+  // This function configures the event handlers of media recorder.
+  monitorRecorderEvents: function(event) {
+    console.log('monitor debug:', event);
```
Comment (KaptenJansson) [inline src/manual/media-recorder/js/main.js]: Remove debug logging.

File: src/manual/media-recorder/js/main.js [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+
+  // This function builds the basic gUM constraints from the
+  // selected GetUserMedia options in html.
+  buildConstraints: function(audioEnabled, videoEnabled, resolution) {
+    var constraints = {audio: false, video: false};
+    var res = resolution.split('x');
+    var width = res[0];
+    var height = res[1];
+    if (height && videoEnabled) {
+      constraints.video = {mandatory: {
```
Comment (KaptenJansson) [inline src/manual/media-recorder/js/main.js]: This is the old constraints style, see this [sample page](https://github.com/webrtc/samples/blob/gh-pages/src/content/getusermedia/resolution/js/main.js) for what the new ones look like.

File: src/manual/media-recorder/js/main.js [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+
+  // This function performs a gUM request and populates the local stream as
+  // well as the localVideo tag.
+  getUserMedia: function() {
+    var enableAudio = this.isTrue(this.getSelectedValue('enableAudio'));
+    var enableVideo = this.isTrue(this.getSelectedValue('enableVideo'));
+    var resolution = this.getSelectedValue('resolution');
+    var constraints = this.buildConstraints(enableAudio, enableVideo,
+        resolution);
+    var me = this;
```
Comment (KaptenJansson) [inline src/manual/media-recorder/js/main.js]: Replace `me` with `self`.

File: src/manual/media-recorder/js/main.js [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+  // This function stops a either a whole stream, or video tracks or audio
+  // tracks from that stream.
+  stopStream: function(stream, tracksType) {
+    return new Promise(function(resolve, reject) {
+      var tracks = null;
+      if (!stream) {
+        alert('Stream is not initialized!');
+        reject();
+      }
+      if (tracksType === 'video') {
```
Comment (KaptenJansson) [inline src/manual/media-recorder/js/main.js]: Why not just do getTracks() and stop all tracks?

File: src/manual/media-recorder/js/main.js [hunk 7]
```diff
@@ -0,0 +1,441 @@
...
+*
+*  Use of this source code is governed by a BSD-style license
+*  that can be found in the LICENSE file in the root of the source
+*  tree.
+*/
+
+'use strict';
+
+function RecorderTest() {
+  this.theLocalStream = null;
```
Comment (KaptenJansson) [inline src/manual/media-recorder/js/main.js]: Omit unnecessary `the` prefix.

File: src/manual/media-recorder/js/main.js [hunk 8]
```diff
@@ -0,0 +1,441 @@
...
+  this.theRemoteStream = null;
```
Comment (KaptenJansson) [inline src/manual/media-recorder/js/main.js]: Omit unnecessary `the` prefix.

File: src/manual/media-recorder/js/main.js [hunk 9]
```diff
@@ -0,0 +1,441 @@
...
+  this.theRecorder = null;
```
Comment (KaptenJansson) [inline src/manual/media-recorder/js/main.js]: Omit unnecessary `the` prefix.

File: src/manual/media-recorder/js/main.js [hunk 10]
```diff
@@ -0,0 +1,441 @@
...
+    var audioBitrate = this.getSelectedValue('audioBitrate');
+    var videoCodec = this.getSelectedValue('videoCodec');
+    var videoBitrate = this.getSelectedValue('videoBitrate');
+    var overallBitrate = this.getSelectedValue('overallAvBitrate');
+    if (stream === 'local') {
+      stream = this.theLocalStream;
+    } else {
+      stream = this.theRemoteStream;
+    }
+    if (!stream) {
```
Comment (KaptenJansson) [inline src/manual/media-recorder/js/main.js]: You could do:

```
if (!stream)  {
  return null
} else if (stream === 'local')  {
  stream = this.theLocalStream;
} else {
  stream = this.theRemoteStream;
}
```

File: src/manual/media-recorder/js/main.js [hunk 11]
```diff
@@ -0,0 +1,441 @@
...
+    console.log('recorder config', config);
+    var ON_DATA_AVAILABLE = 'ondataavailable';
+    if (!config) {
+      alert('Cannot create recorder without a stream!');
+      return;
+    }
+    var stream = config[0];
+    var options = config[1];
+    var ignoreMutedMedia = this.isTrue(this.getSelectedValue('ignoreMuted'));
+    var me = this;
```
Comment (KaptenJansson) [inline src/manual/media-recorder/js/main.js]: Replace me with self.

File: src/manual/media-recorder/js/main.js [hunk 12]
```diff
@@ -0,0 +1,441 @@
...
+  },
+
+  // This function creates the local peer connection and saves the
+  // remote stream object.
+  createPeerConnection: function(stream) {
+    if (!this.theLocalStream) {
+      alert('No local stream initialized!');
+      return;
+    }
+    var me = this;
```
Comment (KaptenJansson) [inline src/manual/media-recorder/js/main.js]: Replace me with self.

File: src/manual/media-recorder/js/main.js [hunk 13]
```diff
@@ -0,0 +1,441 @@
...
+    downloader.click();
+    setTimeout(function() {
+      document.body.removeChild(downloader);
+      window.URL.revokeObjectURL(url);
+    }, 100);
+  },
+
+  // This function creates the local peer connection and saves the
+  // remote stream object.
+  createPeerConnection: function(stream) {
```
Comment (KaptenJansson) [inline src/manual/media-recorder/js/main.js]: Remove `stream` since you never provide a stream.

File: src/manual/media-recorder/js/main.js [hunk 14]
```diff
@@ -0,0 +1,441 @@
+/*
+*  Copyright (c) 2016 The WebRTC project authors. All Rights Reserved.
+*
+*  Use of this source code is governed by a BSD-style license
+*  that can be found in the LICENSE file in the root of the source
+*  tree.
+*/
+
```
Comment (KaptenJansson) [inline src/manual/media-recorder/js/main.js]: Add `/* exported varNames */` for all the variables called upon elsewhere (window.onerror(lineno..), resolve, reject, descrption etc).

Add `/* globals varNames */` for all the variables defined elsewhere (MediaRecorder, Promise)

**Comment by KaptenJansson** [comment]
Also make sure to use the email address used for the CLA in your commits. If you have just signed it, reply here with `I signed it` so that the Google CLA bot can check it.

**Comment by KaptenJansson** [comment]
Ping

**Comment by cricdecyan** [comment]
Hi Chris,
Sorry for the delay, it has been really busy here. I will try to work on it
in a couple weeks when things calm down.
Thanks
Christian

**Comment by KaptenJansson** [comment]
Awesome, thanks for letting me know!

**Comment by KaptenJansson** [comment]
Closing for now.

**Pull request closed without merging**